aws-http-client-url-connection/src/main/java/software/amazon/awssdk/http/java/UrlConnectionHttpClient.java

```diff
@@ -0,0 +1,155 @@
+import software.amazon.awssdk.http.SdkHttpFullRequest;
+import software.amazon.awssdk.http.SdkHttpFullResponse;
+import software.amazon.awssdk.http.SdkRequestContext;
+import software.amazon.awssdk.utils.AttributeMap;
+import software.amazon.awssdk.utils.IoUtils;
+
+@SdkInternalApi
+final class UrlConnectionHttpClient implements SdkHttpClient {
+
+    private final AttributeMap options;
+
+    UrlConnectionHttpClient(AttributeMap options) {
+        this.options = options;
+    }
+
+    @Override
+    public AbortableCallable<SdkHttpFullResponse> prepareRequest(SdkHttpFullRequest request, SdkRequestContext requestContext) {
+        final HttpURLConnection connection = createAndConfigureConnection(request);
+        return new RequestCallable(connection, request);
+    }
+
+    @Override
+    public <T> Optional<T> getConfigurationValue(SdkHttpConfigurationOption<T> key) {
+        return Optional.ofNullable(options.get(key));
+    }
+
+    @Override
+    public void close() throws Exception {
+
+    }
+
+    private HttpURLConnection createAndConfigureConnection(SdkHttpFullRequest request) {
+        HttpURLConnection connection = invokeSafely(() -> (HttpURLConnection) createRequest(request).toURL().openConnection());
+        request.getHeaders().forEach((key, values) -> values.forEach(value -> connection.setRequestProperty(key, value)));
+        invokeSafely(() -> connection.setRequestMethod(request.getHttpMethod().name()));
+        if (request.getContent() != null) {
+            connection.setDoOutput(true);
+        }
+
+        connection.setConnectTimeout(saturatedCast(options.get(CONNECTION_TIMEOUT).toMillis()));
+        connection.setReadTimeout(saturatedCast(options.get(SOCKET_TIMEOUT).toMillis()));
+
+        return connection;
+    }
+
+    private URI createRequest(SdkHttpFullRequest request) {
+        StringBuilder uriBuilder = new StringBuilder(request.getEndpoint().toString());
+        if (isNotBlank(request.getResourcePath())) {
+            uriBuilder.append(request.getResourcePath());
+        }
+
+        String params = request.getParameters().entrySet().stream()
+                               .flatMap(this::flattenParams)
+                               .collect(Collectors.joining("&"));
+
+        if (isNotBlank(params)) {
+            uriBuilder.append("?").append(params);
+        }
+
+        return invokeSafely(() -> new URI(uriBuilder.toString()));
+    }
+
+    private Stream<String> flattenParams(Map.Entry<String, List<String>> e) {
+        if (e.getValue() == null || e.getValue().size() == 0 || e.getValue().get(0) == null) {
+            return Stream.of(encode(e.getKey()));
+        }
+
+        return e.getValue().stream().map(v -> encode(e.getKey()) + "=" + encode(v));
+    }
+
+    private static <K, V> Collector<Map.Entry<K, V>, ?, Map<K, V>> toMap() {
+        return Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue);
+    }
+
+    private static String encode(String string) {
```

Comment:
Most decent HTTP implementations expose a high level request objects that separates those things and handles encoding automatically. If we try and handle that in the SDK those implementations will have to undecode and pull apart those things just to do it again.